### PR TITLE
Add support for StackExchange.Redis

### LINKED ISSUE #874: Redis support
I need to have Redis commands captured in the APM data. I saw somewhere that it is planned, but not currently available. Is this something you would be interested in taking a PR for? If so, do you have any current thoughts on how this would be implemented or any other information you can point me in the right direction on?

### PR DESCRIPTION
**Pull request by russcam**
This commit adds support for tracing redis commands
sent with StackExchange.Redis.

The implementation registers a profiler with
IConnectionMultiplexer that provides a profiling session
per transaction. Added an Ended event to transaction
that a delegate is registered with to end the profiling session and
create a span for each profiled command.

Added StackExchangeRedisSample to demonstrate how to register
Elastic APM agent. For ASP.NET Core, registration is performed in
`ConfigureServices(IServiceCollection services)`:

```c#
services.AddSingleton<IConnectionMultiplexer>(provider =>
{
	var connection = ConnectionMultiplexer.Connect("localhost");
	connection.UseElasticApm();
	return connection;
});
```

Closes #874

### CODE REVIEW

File: src/Elastic.Apm/Model/Transaction.cs [hunk 1]
```diff
@@ -378,6 +383,9 @@ public void End()
 			_isEnded = true;
 			if (!isFirstEndCall) return;
 
+			var handler = Ended;
+			handler?.Invoke(this, EventArgs.Empty);
```
Comment (gregkalapos) [inline src/Elastic.Apm/Model/Transaction.cs:387]: Would it make sense to do `Ended = null` here? Just to make sure we don't hold a reference to anything longer than needed.

**Review by gregkalapos — APPROVED** [review]
Nice! We have a discussion on slack regarding parent spans for the generated redis spans. I think the current version is already a good start and definitely fine to merge, no reason to block this. And then we can do follow up PRs if needed.

Also added 2 smaller comments.

File: src/Elastic.Apm.StackExchange.Redis/ElasticApmProfiler.cs [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+				// update the timestamp to reflect the point at which the command was created
+				if (span is Span realSpan)
+					realSpan.Timestamp = TimeUtils.ToTimestamp(profiledCommand.CommandCreated);
+
+				span.Duration = profiledCommand.ElapsedTime.TotalMilliseconds;
+
+				// profiled commands are always successful
+				span.Outcome = Outcome.Success;
+
+				// TODO: clear the raw stacktrace as it won't be representative of the call stack at
```
[the remark this answers is not in the document]
Comment (russcam) [reply src/Elastic.Apm.StackExchange.Redis/ElasticApmProfiler.cs:141]: I'm leaning towards having an argument passed to span to not collect the stacktrace. If we were to capture the stacktraces at the point of `GetProfilingSession()` being called, we would need correlate the stacktraces back to commands captured in the session, as you suggest. I do wonder how much value these stacktraces would have in practice.

Happy for me to following this up in another commit?

File: src/Elastic.Apm/Model/Transaction.cs (hunk 1 above)
Comment (russcam) [reply src/Elastic.Apm/Model/Transaction.cs:387]: yep, will add.

**Comment by russcam** [comment]
I've updated to create a new profiling session per current span or transaction, so that redis command spans will be captured as sub spans of both spans and transactions.

**Comment by russcam** [comment]
jenkins please run the tests

**Comment by russcam** [comment]
Looks like Linux tests timed out, though all that ran passed.

File: src/Elastic.Apm.StackExchange.Redis/ElasticApmProfiler.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (gregkalapos) [reply src/Elastic.Apm.StackExchange.Redis/ElasticApmProfiler.cs:173]: Agreed on not trying to capture stack traces at this point.

Sure - let's do this outside this PR.

**Pull request merged**